messaging/src/main/java/org/axonframework/configuration/DefaultAxonApplication.java

```diff
@@ -58,123 +62,125 @@ class DefaultAxonApplication extends AbstractConfigurer<AxonApplication> impleme
 
     private final TreeMap<Integer, List<LifecycleHandler>> startHandlers = new TreeMap<>();
     private final TreeMap<Integer, List<LifecycleHandler>> shutdownHandlers = new TreeMap<>(Comparator.reverseOrder());
+    private final DefaultComponentRegistry componentRegistry;
     private long lifecyclePhaseTimeout = 5;
     private TimeUnit lifecyclePhaseTimeunit = TimeUnit.SECONDS;
     private boolean enhancerScanning = true;
 
-    private final AxonConfigurationImpl axonConfig = new AxonConfigurationImpl();
-    private final AtomicBoolean initialized = new AtomicBoolean(false);
+    private final AtomicReference<AxonConfiguration> configuration = new AtomicReference<>();
+
+    public DefaultAxonApplication() {
+        this.componentRegistry = new DefaultComponentRegistry();
+    }
 
     @Override
-    public void onStart(int phase, @Nonnull LifecycleHandler startHandler) {
-        registerLifecycleHandler(startHandlers, phase, startHandler);
+    public DefaultAxonApplication onStart(int phase, @Nonnull LifecycleHandler startHandler) {
+        return registerLifecycleHandler(startHandlers, phase, startHandler);
     }
 
     @Override
-    public void onShutdown(int phase, @Nonnull LifecycleHandler shutdownHandler) {
-        registerLifecycleHandler(shutdownHandlers, phase, shutdownHandler);
+    public DefaultAxonApplication onShutdown(int phase, @Nonnull LifecycleHandler shutdownHandler) {
+        return registerLifecycleHandler(shutdownHandlers, phase, shutdownHandler);
     }
 
-    private void registerLifecycleHandler(Map<Integer, List<LifecycleHandler>> lifecycleHandlers,
+    private DefaultAxonApplication registerLifecycleHandler(Map<Integer, List<LifecycleHandler>> lifecycleHandlers,
                                           int phase,
                                           @Nonnull LifecycleHandler lifecycleHandler) {
+        if (configuration.get() != null) {
+            throw new IllegalArgumentException(
+                    "Cannot register lifecycle handlers when the configuration is already initialized");
+        }
         lifecycleHandlers.compute(phase, (p, handlers) -> {
             if (handlers == null) {
                 handlers = new CopyOnWriteArrayList<>();
             }
             handlers.add(requireNonNull(lifecycleHandler, "Cannot register null lifecycle handlers."));
             return handlers;
         });
+        return this;
     }
 
     @Override
-    public AxonApplication registerLifecyclePhaseTimeout(long timeout, @Nonnull TimeUnit timeUnit) {
+    public DefaultAxonApplication registerLifecyclePhaseTimeout(long timeout, @Nonnull TimeUnit timeUnit) {
```

Comment:
Ow here is the life cycle stuff! :heart: 
Shouldn't this method also reside on an interface somewhere, for documentation purposes?